### PR TITLE
Create Rhinoceros adapter

### LINKED ISSUE #199: Scrap multiple models for all of their data
#### Description:

For when sent multiple models from an external collaborator, it can be useful to merge the models into one.

First step of doing this is to automatically open a list of models, and output all of their data.

For merging the model, we can then to start with use Human or similar to bake the geometry out into a fresh model.

### PR DESCRIPTION
**Pull request by rolyhudson**
### Issues addressed by this PR
Rudimentary adapter set up for very basic pull and push.

Closes #199 

### Test files
[Simple pull and push](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Rhinoceros_Toolkit/%23205-CreateRhinocerosAdapter?csf=1&web=1&e=HhUOVI)

### CODE REVIEW

File: Rhinoceros_oM/BHoMRhinoObject.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public class BHoMRhinoObject : BHoMObject
+    {
+        /*******************************************/
+        /**** Properties                        ****/
+        /*******************************************/
+        public virtual string Layer { get; set; } = "";
+
+        public virtual IGeometry Geometry { get; set; } = null;
+
+        public virtual Color LayerColour { get; set; } = new Color();
```
Comment (IsakNaslundBh) [inline Rhinoceros_oM/BHoMRhinoObject.cs:21]: Would consider adding a `Layer` class as well to handle the layer colour and name. Then this object could just store a ref to that.

File: Rhinoceros_Adapter/RhinocerosAdapter.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            }
+
+
+
+            m_RhinoceroSettings = fileSettings;
+
+        }
+
+        private static BH.oM.Adapter.FileSettings m_RhinoceroSettings { get; set; } = null;
+        private static List<string> m_FilePaths { get; set; } = new List<string>();
```
Comment (IsakNaslundBh) [inline Rhinoceros_Adapter/RhinocerosAdapter.cs:56]: Not sure I think this should be static? Should there only be one list, even if you have multiple adapters?

```suggestion
        private List<string> m_FilePaths { get; set; } = new List<string>();
```

File: Rhinoceros_Adapter/RhinocerosAdapter.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                return;
+            }
+
+
+
+            m_RhinoceroSettings = fileSettings;
+
+        }
+
+        private static BH.oM.Adapter.FileSettings m_RhinoceroSettings { get; set; } = null;
```
Comment (IsakNaslundBh) [inline Rhinoceros_Adapter/RhinocerosAdapter.cs:55]:
```suggestion
        private BH.oM.Adapter.FileSettings m_RhinoceroSettings { get; set; } = null;
```

Same for this, should it really be static? Also, is there any harm in making this public even?

```suggestion
        public virtual BH.oM.Adapter.FileSettings RhinoceroSettings { get; set; } = null;
```

File: Rhinoceros_Adapter/AdapterActions/Push.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+using BH.oM.Data.Requests;
+using BH.oM.Adapter;
+using BH.oM.Base;
+using System.Reflection;
+
+namespace BH.Adapter.Rhinoceros
+{
+    public partial class RhinocerosAdapter : BHoMAdapter
+    {
+        public override List<object> Push(IEnumerable<object> objects, String tag = "", PushType pushType = PushType.AdapterDefault, ActionConfig actionConfig = null)
```
Comment (IsakNaslundBh) [inline Rhinoceros_Adapter/AdapterActions/Push.cs:38]: Not sure we really should be overriding this? Or, might be me missing the need?

File: Rhinoceros_Adapter/AdapterActions/Pull.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+using BH.oM.Adapter;
+using BH.oM.Base;
+using BH.Engine.Adapter;
+using System.IO;
+
+namespace BH.Adapter.Rhinoceros
+{
+    public partial class RhinocerosAdapter : BHoMAdapter
+    {
+        public override IEnumerable<object> Pull(IRequest request, PullType pullType = PullType.AdapterDefault, ActionConfig actionConfig = null)
```
Comment (IsakNaslundBh) [inline Rhinoceros_Adapter/AdapterActions/Pull.cs:39]: Same for this, do not think we should need to override this? I think the filter request would call the same thing anyway

**Review by IsakNaslundBh — COMMENTED** [review]
Have not tested, just quickly scrolled through the code here with some initial comments.

Have not made any deep dive, so please do _not_ take the comments as a "I am sure this is better if changed" but rather as question/suggestion.

Will be quite nice to get this in for sure!

File: Rhinoceros_Adapter/AdapterActions/Push.cs (hunk 4 above)
Comment (rolyhudson) [reply Rhinoceros_Adapter/AdapterActions/Push.cs:38]: Good questions. I am just following a pattern of override use on Push and Pull observed in other Adapters. I can't answer why though!

File: Rhinoceros_Adapter/AdapterActions/Pull.cs (hunk 5 above)
Comment (rolyhudson) [reply Rhinoceros_Adapter/AdapterActions/Pull.cs:39]: See above.

File: Rhinoceros_Adapter/AdapterActions/Push.cs (hunk 4 above)
Comment (IsakNaslundBh) [reply Rhinoceros_Adapter/AdapterActions/Push.cs:38]: ok, think those cases should be seen as exceptions in general! If it works without overriding, I think we should skip overriding! :)

Comment (rolyhudson) [reply Rhinoceros_Adapter/AdapterActions/Push.cs:38]: I cannot find an adapter where Push and Pull are not overridden! Maybe @alelom can clarify why this is set up in this way?

File: Rhinoceros_oM/BHoMRhinoObject.cs (hunk 1 above)
Comment (rolyhudson) [reply Rhinoceros_oM/BHoMRhinoObject.cs]: Good idea. Added this.

File: Rhinoceros_Adapter/AdapterActions/Push.cs (hunk 4 above)
Comment (IsakNaslundBh) [reply Rhinoceros_Adapter/AdapterActions/Push.cs:38]: Non of the structural adapters override it, or at least, any override is only made for the sake of doing some initial checks on the adapter or so, before the `base.Push()` is called :)

**Comment by rolyhudson** [comment]
Refactored this now after testing approaches for very basic workflows. 
Anticipating further development to handle requests and more complex CRUD operations. 
This PR aims to capture the read workflow indicated in #199 plus a simple create file workflow.

For the Pull Action this means simply read everything from the specified file. 
Objects are stored as the `BH.oM.Adapters.Rhinoceros.RhinoObject` capturing geometry, object colour and layer.

For the Push Action the initial workflow is seen as simply writing to a new file or overwriting if an existing file is specified.  
To do so Push is overridden. 
Why? 

1. This provides the mechanism to create a single 3dm file object where objects added.
2. To enforce that this adapter (for now) is `PushType.CreateOnly` and expose a warning for users that this is the case.

Would be good to get @IsakNaslundBh's view on this.

**Review by alelom — APPROVED** [review]
As a prototype I'm more than happy with this.

We can then work and improve. 

I think we can remove the Push override, because currently it is done to expose a relevant error message if something different than `PushType.CreateOnly` is used, enforcing it. The same can actually be achieved by overriding `SetupPushType()` (should've mentioned it before, sorry).

Anyway, for me this is good to merge at this current state. @rolyhudson happy to discuss and raise together relevant issues to improve this.

**Comment by FraserGreenroyd** [comment]
@rolyhudson and @alelom (and @IsakNaslundBh to a lesser extent). Given this PR is described as a `rudimentary adapter set up for very basic pull and push` and this wasn't available in the 4.0 beta, can I check we are happy with the use cases/risks vs reward of merging this in now for inclusion in the 4.1 beta, vs waiting for a merge in the first sprint of 4.2, getting some alpha testing under its belt for inclusion in the 4.2 beta.

Not wanting to hold things up unnecessarily, just raising the question for discussion to ensure we're happy that this work will be deployed in installers globally in 2 weeks without much opportunity for alpha users to feedback :smile:

**Comment by IsakNaslundBh** [comment]
> Given this PR is described as a rudimentary adapter set up for very basic pull and push and this wasn't available in the 4.0 beta, can I check we are happy with the use cases/risks vs reward of merging this in now for inclusion in the 4.1 beta, vs waiting for a merge in the first sprint of 4.2, getting some alpha testing under its belt for inclusion in the 4.2 beta.

From my point of view, I think it is fine to wait, unless @rolyhudson and @alelom disagrees due to needs for some workflows.

**Comment by rolyhudson** [comment]
I am happy to wait and test fully in 4.2 alpha's. For me it was an opportunity to look deeper into adapters so no need for haste in merging.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Pull request closed without merging**